client/my-sites/media-library/media-date-range.jsx

```diff
@@ -0,0 +1,344 @@
+		// iii) before today (don't allow inputs to select dates that calendar is not allowed to show)
+		const isValidFrom =
+			fromDate.isValid() && fromDate.isSameOrAfter( epoch ) && fromDate.isSameOrBefore( today );
+
+		const isValidTo =
+			toDate.isValid() && toDate.isSameOrAfter( epoch ) && toDate.isSameOrBefore( today );
+
+		// If either of the date inputs are invalid then revert
+		// to current start/end date from state
+		if ( ! isValidFrom || ! isValidTo ) {
+			this.setState( {
+				inputFromDate: this.dateToHumanReadable( this.state.startDate ),
+				inputToDate: this.dateToHumanReadable( this.state.endDate ),
+			} );
+		}
+
+		// If the new date in the blurred input is valid...
+		// ...and it's
+		if (
+			date.isValid() &&
+			date.isSameOrAfter( epoch ) &&
+			date.isSameOrBefore( today ) &&
+			! this.state[ startOrEnd ].isSame( date, 'day' )
+		) {
+			this.onSelectDate( date );
+		}
+	}
+
+	onSelectDate( date ) {
+		// DateUtils requires a range object with this shape
+		const range = {
+			from: this.momentDateToNative( this.state.startDate ),
+			to: this.momentDateToNative( this.state.endDate ),
+		};
+
+		const rawDay = this.momentDateToNative( date );
+
+		// Calculate the new Date range
+		const newRange = DateUtils.addDayToRange( rawDay, range );
+
+		// Edge case: Range can sometimes be from: null, to: null
+		if ( ! newRange.from || ! newRange.to ) return;
+
+		this.setState(
+			previousState => {
+				let newState = {
+					startDate: this.nativeDateToMoment( newRange.from ),
+					endDate: this.nativeDateToMoment( newRange.to ),
+					inputFromDate: this.nativeDateToMoment( newRange.from ).format( 'L' ),
+					inputToDate: this.nativeDateToMoment( newRange.to ).format( 'L' ),
+				};
+
+				// For first date selection only: "cache" previous dates
+				// just in case user doesn't "Apply" and we need to revert
+				// to the original dates
+				if ( ! this.state.oldDatesSaved ) {
+					newState = Object.assign( {}, newState, {
+						oldStartDate: previousState.startDate,
+						oldEndDate: previousState.endDate,
+						oldDatesSaved: true, // marks that we have saved old dates
+					} );
+				}
+
+				return newState;
+			},
+			() => this.props.onDateSelect( this.state.startDate, this.state.endDate )
+		);
+	}
+
+	revertDates() {
+		this.setState( previousState => {
+			const newState = { oldDatesSaved: false };
+
+			if ( previousState.oldStartDate && previousState.oldEndDate ) {
+				newState.startDate = previousState.oldStartDate;
+				newState.endDate = previousState.oldEndDate;
+			}
+
+			return newState;
+		} );
+	}
+
+	commitDates() {
+		this.setState(
+			previousState => ( {
+				oldStartDate: previousState.startDate, // update cached old dates
+				oldEndDate: previousState.endDate, // update cached old dates
+				oldDatesSaved: false,
+			} ),
+			() => {
+				this.props.onDateCommit( this.state.startDate, this.state.endDate );
+				this.togglePopover();
```

Comment:
I think we would better show the intent to be calling a `closePopover` method here right? Is there a case where we would call `commitDates` and expect `togglePopover` to open the popover?

Reply:
I agree we should be making the intent obvious. If I were writing this now I'd have 2 methods for `open` and `close` which proxied the correct args to the master `toggle`. I hate the ambiguity of `toggle`.

> Is there a case where we would call commitDates and expect togglePopover to open the popover?

I feel like I wouldn't have done this for no reason but then again I was dipping in and out of this code so it's entirely possible I just made a lazy mistake. Will investigate.

Reply:
Created dedicated methods in https://github.com/Automattic/wp-calypso/pull/29938 for `open` and `close`.